identity-service/src/routes/userBankTransactionMetadata.js

```diff
@@ -0,0 +1,94 @@
+const models = require("../models");
+const {
+  handleResponse,
+  successResponse,
+  errorResponseBadRequest,
+  errorResponseServerError,
+} = require("../apiHelpers");
+const { logger } = require("../logging");
+const { Op } = require("sequelize");
+
+const requiredPurchaseKeys = {
+  discriminator: { type: "string" },
+  usd: { type: "string" },
+  sol: { type: "string" },
+  audio: { type: "string" },
+  swapTransaction: { type: "string" },
+  buyTransaction: { type: "string" },
+};
+
+const validateMetadata = (metadata) => {
+  if (!metadata) {
+    throw new Error("Missing metadata");
+  }
+  if (!("discriminator" in metadata)) {
+    throw new Error(`Missing discriminator`);
+  }
+  if (metadata.discriminator === "PURCHASE_SOL_AUDIO_SWAP") {
+    const result = {};
+    for (let key of Object.keys(requiredPurchaseKeys)) {
+      if (!(key in metadata)) {
+        throw new Error(`Missing required property '${key}'`);
+      }
+      if (typeof metadata[key] !== requiredPurchaseKeys[key].type) {
+        throw new Error(
+          `Invalid type for property '${key}'. Found '${typeof metadata[
+            key
+          ]}', expected '${requiredPurchaseKeys[key].type}'`
+        );
+      }
+      result[key] = metadata[key];
+    }
+    return result;
+  }
+  throw new Error("Invalid discriminator");
+};
+
+module.exports = function (app) {
+  app.post(
+    "/transaction_metadata",
+    handleResponse(async (req, res, next) => {
```

Comment:
We should definitely authenticate this endpoint, and probably want to include user_id in the schema somewhere so somebody else can't come and overwrite your TX history with bogus data

Reply:
makes sense